### ISSUE TITLE
LWS_CALLBACK_CLOSED_HTTP not triggered on connection timeout

### ISSUE THREAD

**Issue by dev31337** [warmcat/libwebsockets#1923]
Hi,

If you make a connection to the http server (for example using netcat) and wait for the timeout (10s) to happen, the event LWS_CALLBACK_CLOSED_HTTP is not triggered.

I think it is not intended because this event is important to track disconnections. 

Regards.

**Comment by lws-team** [comment]
The different protocols, and if the protocol has the concept, whether it's client or server side, have different sets of callbacks.  You want LWS_CALLBACK_CLOSED_CLIENT_HTTP

https://libwebsockets.org/git/libwebsockets/tree/minimal-examples/http-client/minimal-http-client/minimal-http-client.c#n127

**Comment by dev31337** [comment]
As I was telling in my post this is a HTTP server (not a client) and thus LWS_CALLBACK_CLOSED_CLIENT_HTTP is not triggered (because it's for client only).

But I strongly think that LWS_CALLBACK_CLOSED_HTTP should be triggered on incoming server connection timeout.

Note: I track all the events and on incoming connection timeout only LWS_CALLBACK_WSI_DESTROY is triggered.

**Comment by dev31337** [comment]
Sorry but why did you close this issue ?

May be I did not explain it clearly.

The issue is with the HTTP server protocol (not the client).

If lws receives an incoming HTTP connection and that connection times out (because it does not receive any data) then LWS_CALLBACK_CLOSED_HTTP is not triggered.

So there is no way for the application to know that the connection was closed due to inactivity.

**Comment by lws-team** [comment]
>> If you make a connection to the http server 

Sorry that read to me like we're the client doing that.

The connection lifecycle is not always related to http1.1 transaction (and thus protocol) type.  There are different, inner callbacks designed to be used as the place to allocate and deallocate

https://libwebsockets.org/git/libwebsockets/tree/include/libwebsockets/lws-callbacks.h#n310-326

**Comment by dev31337** [comment]
I think I understand your point, you mean that an inactive incoming connection is not considered as an HTTP connection.

But here is the issue that I have, I'm making a connection filter to avoid DDOS attacks to libwesockets, I need to know when there is an incoming TCP connection and when that incoming connection is closed.

I know about an incoming connection using LWS_CALLBACK_FILTER_NETWORK_CONNECTION (and I know its IP), but I have no way to know when that connection is closed by timeout, the only event triggered in that case is LWS_CALLBACK_WSI_DESTROY and the wsi is not readable anymore to be able to find its IP and match it to a connection.

**Comment by lws-team** [comment]
For the DDOS filter, yes you would be interested in close.  For everything else the protocol bind is more interesting.

`LWS_WITH_PEER_LIMITS` might already do what you want

https://libwebsockets.org/git/libwebsockets/tree/include/libwebsockets/lws-context-vhost.h#n546-560
https://libwebsockets.org/git/libwebsockets/tree/lib/misc/peer-limits.c

**Comment by dev31337** [comment]
I just tried LWS_WITH_PEER_LIMITS which seems to work but has one issue when using ip_limit_wsi it does not close the extra connections.

I edited lib/core-net/adopt.c:

        if (info->type & LWS_ADOPT_SOCKET) {
                peer = lws_get_or_create_peer(info->vh, info->fd.sockfd);

                if (peer && info->vh->context->ip_limit_wsi &&
                    peer->count_wsi >= info->vh->context->ip_limit_wsi) {
                        lwsl_notice("Peer reached wsi limit %d\n",
                                        info->vh->context->ip_limit_wsi);
                        lws_stats_bump(&info->vh->context->pt[0],
                                              LWSSTATS_C_PEER_LIMIT_WSI_DENIED,
                                              1);
                        compatible_close(info->fd.sockfd);
                        return NULL;
                }

I added "compatible_close(info->fd.sockfd);" which was not here in the original code and now it closes extra connections as it should.

**Comment by dev31337** [comment]
In order to do my own DDOS and not use LWS_WITH_PEER_LIMITS (because I need notification of such DDOS) I also did a small modification to "lib/core-net/close.c.

At this time lws close the fd of the wsi before triggering LWS_CALLBACK_WSI_DESTROY.

I think it's a bad decision because it prevents the application from accessing the remote IP of the connection from that event.

I modified it so the event is triggered before closing the fd.

**Comment by lws-team** [comment]
Thanks for the fix with the close, I pushed a patch on master and v4.0-stable for it.

WSI_DESTROY happens even if there is no socket fd, or it's a file fd as well... the wsi is destroyed right at the end, earlier in the close flow some protocols like ws and h2 may defer the close so it can try to issue a CLOSE frame inside the protocol.  So how it is, is how it wants to be I think.

PEER_LIMITS knows the IP, is there something we can do with a callback you can set in the context info or lws_system to let you see what is happening in there?

**Comment by dev31337** [comment]
Sure, a new callback which would notify the application when the peer limit is reached (and thus the new connection is closed) would be great.

I would be able to get the remote IP from the wsi which triggered it to report/alert about it. (It is very important for security monitoring)

What I like about your LWS_WITH_PEER_LIMITS is the concept of soft and hard limit, this is very smart, I would love to use it.

**Comment by lws-team** [comment]
OK.  I wrote PEER_LIMITS a few years ago and it has missed out on some core improvements since then like the change in the rest of lws to lws_sockaddr46 to handle ipv4 or ipv6; it supports it but does it by hand.  So it needs some cleaning and updating, I will add a callback in the context info as part of that.

**Comment by lws-team** [comment]
I modernized it and added the callback, please take a look on master.  It's only build-tested but the changes are pretty straightforward.

**Comment by dev31337** [comment]
Tanks it's working but I get an error message I did not have with my version 4.0.1:

lws_server_socket_service_ssl: PEEKed 0 (from_pollin 1)

This happens when I netcat to the server port then disconnects it. (without sending any data, it happens at the disconnection)

**Comment by dev31337** [comment]
I would like also to report an inconsistency (which troubles my code), when you netcat to the server port and then disconnects you get the event LWS_CALLBACK_SERVER_NEW_CLIENT_INSTANTIATED (at connection) and LWS_CALLBACK_CLOSED_HTTP (at disconnection) which is perfect.

Unfortunately if you netcat to the server port and wait for the timeout (10s) you only get the event LWS_CALLBACK_SERVER_NEW_CLIENT_INSTANTIATED (at the connection) but not the event LWS_CALLBACK_CLOSED_HTTP at the timeout.

Do you think you could add/trigger it for consistency ? If not can you please direct me to where I could do/insert it myself (because I need it for my code).

**Comment by dev31337** [comment]
In order to explain you why I'm going through these requests/questions, I'm working on security with libwesockets and I need to see how I can manage (detect/report/handle) port scanners (that open connections, send no data and disconnect) and DDOS attacks which opens as many connections as possible.

**Comment by dev31337** [comment]
There is also another error that has been introduced along the line (form my 4.01), I now get the error message "lws_context_destroy3" when closing/shutting down.

You can even see this error message output at this URL:

https://github.com/warmcat/libwebsockets/tree/master/minimal-examples/secure-streams/minimal-secure-streams-staticpolicy

In the README.md you will see the error line "[2020/03/26 15:49:14:2265] E: lws_context_destroy3".

**Comment by lws-team** [comment]
> you changed the message from lwsl_debug to lwsl_warn 
>  another error that has been introduced

Thee are just leftovers from debugging expeditions on my part.  I increased the log level or added something like `lwsl_err("%s", __func__);` because I am looking at a very verbose log spew with thousands of lines, and at that time I need to quickly know when those things happen in the logs so I can figure out what is happening.  It doesn't indicate any error or something to warn about and just a failure to clean up my "debug by printf" after the bug is fixed, often because I got completely exhausted forcing it to reveal itself and solving it.

I made a patch to clean those up and pushed it on master, those logs are already OK on v4.0-stable.

> I'm working on security with libwebsockets

I run lws for all my serving on the public internet for some years, I am also invested in it being pretty secure.  It is tested by Coverity and that has caught a lot of potential problems too.  It builds with -Werror -Wall.  "Security" means a lot of different things, but at least these are "green flags".

> inconsistency

I can't reproduce this using minimal-http-server and this patch to observe what it's doing (here is a case of the same 'debug by printf')

```
diff --git a/lib/core-net/close.c b/lib/core-net/close.c
index 9f4d09206..79b9594d7 100644
--- a/lib/core-net/close.c
+++ b/lib/core-net/close.c
@@ -607,10 +607,12 @@ just_kill_connection:
                if (!wsi->protocol && wsi->vhost && wsi->vhost->protocols)
                        pro = &wsi->vhost->protocols[0];
 
-               if (pro)
+               if (pro) {
+                       lwsl_err("%s: close cb role %s, pro %s\n", __func__, wsi->role_ops->name, pro->name);
                        pro->callback(wsi,
                                wsi->role_ops->close_cb[lwsi_role_server(wsi)],
                                wsi->user_space, NULL, 0);
+               }
                wsi->told_user_closed = 1;
        }
 
```
I am getting

[2020/05/25 06:01:39:8904] E: __lws_close_free_wsi: close cb role h1, pro http-only

either way

**Comment by dev31337** [comment]
I was able to find the issue but I'm stuck in debugging it further you may help.

The issue comes from this line:

        if ((lwsi_state_est_PRE_CLOSE(wsi) ||
            /* raw skt adopted but didn't complete tls hs should CLOSE */
            (wsi->role_ops == &role_ops_raw_skt && !lwsi_role_client(wsi)) ||
             lwsi_state_PRE_CLOSE(wsi) == LRS_WAITING_SERVER_REPLY) &&
            !wsi->told_user_closed &&
            wsi->role_ops->close_cb[lwsi_role_server(wsi)]) {
                if (!wsi->upgraded_to_http2 || !lwsi_role_client(wsi))
                        ccb = 1;

When the connection is closed by the peer lwsi_state_est_PRE_CLOSE(wsi) returns 1 and then ccb is set to 1 and the callback is done.

When the connection is closed by lws because of the timeout lwsi_state_est_PRE_CLOSE(wsi) returns 0 and then ccb is set to 0 and the callback is not done.

Can you please help a bit about it ?

**Comment by lws-team** [comment]
I don't mind looking at it if you can describe how to reproduce it using the minimal examples.  It seems relevant when I tried I didn't see the problem.

**Comment by dev31337** [comment]
I think you can not reproduce it because I'm using TLS and the example is not.

I was able to dig further.

When the remote connection is closed the flag wsi->wsistate_pre_close is LRS_SHUTDOWN.
When the remote connection times out (and lws closes it) the flag wsi->wsistate_pre_close is (LWSIFS_NOT_EST | LRS_SSL_ACK_PENDING).

lwsi_state_est_PRE_CLOSE(wsi) checks for LWSIFS_NOT_EST, and when present does not trigger the callback.

**Comment by lws-team** [comment]
The meaning of lwsi_state_est_PRE_CLOSE() is to find out if the connection was "established", in, eg, ws there might be a lot going on but the connection not really "connected" until it upgrades from http to ws successfully.  Client connections that did not get so far as to be "established" report _CONNECTION_ERROR and not _CLOSED.  Only "established" connections "close".

You can try change the first line you pasted above to 

`if (lwsi_role_server(wsi) || lwsi_state_est_PRE_CLOSE(wsi) ||`

**Comment by dev31337** [comment]
I modified the test:

        if ((lwsi_state_est_PRE_CLOSE(wsi) ||
            /* raw skt adopted but didn't complete tls hs should CLOSE */
            (wsi->role_ops == &role_ops_raw_skt && !lwsi_role_client(wsi)) ||

            (lwsi_state_PRE_CLOSE(wsi) == LRS_SSL_ACK_PENDING && !lwsi_role_client(wsi)) ||

             lwsi_state_PRE_CLOSE(wsi) == LRS_WAITING_SERVER_REPLY) &&
            !wsi->told_user_closed &&
            wsi->role_ops->close_cb[lwsi_role_server(wsi)]) {
                if (!wsi->upgraded_to_http2 || !lwsi_role_client(wsi))
                        ccb = 1;

And now it's working as it should :)

**Comment by dev31337** [comment]
I just found out your answer, I will try your answer too and see if it works.

**Comment by lws-team** [comment]
> ... as it should
> (lwsi_state_PRE_CLOSE(wsi) == LRS_SSL_ACK_PENDING && !lwsi_role_client(wsi)) ||

I don't think the server protocol should hear about a CLOSE on a connection that didn't even set up the tls tunnel.... it wouldn't be controversial to say that it shouldn't hear CLOSE if the tcp level connect did not complete... this is basically the same.

**Comment by dev31337** [comment]
Your modification is good and solves the issue so now the new test is:

       if ((lwsi_role_server(wsi) || lwsi_state_est_PRE_CLOSE(wsi) ||
            /* raw skt adopted but didn't complete tls hs should CLOSE */
            (wsi->role_ops == &role_ops_raw_skt && !lwsi_role_client(wsi)) ||
             lwsi_state_PRE_CLOSE(wsi) == LRS_WAITING_SERVER_REPLY) &&
            !wsi->told_user_closed &&
            wsi->role_ops->close_cb[lwsi_role_server(wsi)]) {
                if (!wsi->upgraded_to_http2 || !lwsi_role_client(wsi))
                        ccb = 1;

It it is fine for you :)

**Comment by dev31337** [comment]
> 
> 
> > ... as it should
> > (lwsi_state_PRE_CLOSE(wsi) == LRS_SSL_ACK_PENDING && !lwsi_role_client(wsi)) ||
> 
> I don't think the server protocol should hear about a CLOSE on a connection that didn't even set up the tls tunnel.... it wouldn't be controversial to say that it shouldn't hear CLOSE if the tcp level connect did not complete... this is basically the same.

In an ideal world the event LWS_CALLBACK_SERVER_NEW_CLIENT_INSTANTIATED should have a counterpart (like LWS_CALLBACK_SERVER_CLIENT_CLOSED) but instead LWS_CALLBACK_CLOSED_HTTP is triggered (when the connection disconnects without sending any data).

So I'm using LWS_CALLBACK_CLOSED_HTTP to track down the disconnections.

**Comment by dev31337** [comment]
One additional question in the pl_notify_cb callback, I get the fd and the sa46 (so the IP), is there a fast way to get the hostname from that ? (I can't use lws_get_peer_addresses() because I have no wsi)

Do I have to resolve it on my own ?

**Comment by lws-team** [comment]
You don't want to do any rdns on the event loop, the C apis for it are blocking and an attacker can DDoS you nicely that way too.

There are native sa46 helpers for parse and format numeric addresses for ipv4 / 6

https://libwebsockets.org/git/libwebsockets/tree/include/libwebsockets/lws-network-helper.h#n179-191

You can log those and / or copy the sa46 object on to a list that is resolved in a separate thread with a cache.

**Comment by dev31337** [comment]
Do you know if lws_get_peer_addresses uses a cache itself or does it resolve the IP at each call ?

**Comment by dev31337** [comment]
I found another warning message, I'm not sure if warning is the right level.

lib/tls/tls-server.c:

               case LWS_SSL_CAPABLE_ERROR:
                        lws_stats_bump(pt, LWSSTATS_C_SSL_CONNECTIONS_FAILED, 1);
                        lwsl_warn("%s: SSL_accept failed socket %u: %d\n",
                                        __func__, wsi->desc.sockfd, n);

This warning is triggered each time you connect to a HTTP SSL port, sends no data and disconnects.

What is your input on this ?

**Comment by dev31337** [comment]
You are doing something quite weird in the function "lws_get_peer_simple_fd" @ lib/core-net/network.c:

        if (getpeername(fd, (struct sockaddr *)&sa46, &len) < 0) {
                lws_snprintf(name, namelen, "getpeername: %s",
                                strerror(LWS_ERRNO));
                return name;
        }

You are returning the error in the string which is supposed to contain the IP.

Not only weird but the application does not expect that and has no easy way to know there was an error.

I personally changed "return name;" to "return NULL;" so the application has a way to know that there was an issue.

I encountered the issue because the fd was closed when LWS_CALLBACK_CLOSED_HTTP was triggered and lws_get_peer_simple_fd thus failed. (a very specific DDoS scenario triggered it)
